### ISSUE TITLE
Proposal to prevent ambiguity in variable assignment

### ISSUE THREAD

**Issue by tBuLi** [tBuLi/symfit#30]
In the current API, the way data is assigned to the variables can be unclear. Although symfit check that either the first or the last dimension of the xdata needs to have the same length as the model has vars, this can be quite unreadable.Furthermore, the unpacking is done by the order in which the data is presented in this dimension, so if the user provides them in the wrong order the fit will be wrong.

Example in the current API, conserning some function defined on the grid 5 <= x < 5, 5 <= y < 5:
```python
xdata = np.arange(-5, 5, 1) 
ydata = np.arange(-5, 5, 1) 
xx, yy = np.meshgrid(xdata, ydata, sparse=False) 
xdata_coor = np.dstack((xx, yy)) 

zdata = 2.5 * xx**2 + 3.0 * yy**2

a, b = parameters('a, b')
x, y = variables('x, y')
new = a * x**2 + b * y**2

fit = Fit(new, xdata_coor, zdata) 
```

I would like to change the API so as to prevent these ambiquity and to increase readability.Proposal:

```python
xdata = np.arange(-5, 5, 1) 
ydata = np.arange(-5, 5, 1) 
xx, yy = np.meshgrid(xdata, ydata, sparse=False) 

zdata = 2.5 * xx**2 + 3.0 * yy**2

a, b = parameters('a, b')
x, y = Variable(data=xx), Variable(data=yy) # We have to prepare the objects seperately because we want to supply kwargs
new = a * x**2 + b * y**2

fit = Fit(new, zdata) 
```
This would increase readability, and remove the need for stacking of the data. However, the call to fit is less clear to me.

Alternatively,

```python
xdata = np.arange(-5, 5, 1) 
ydata = np.arange(-5, 5, 1) 
xx, yy = np.meshgrid(xdata, ydata, sparse=False) 

zdata = 2.5 * xx**2 + 3.0 * yy**2

a, b = parameters('a, b')
x, y = variables('x, y')
new = a * x**2 + b * y**2

fit = Fit(new, x=xx, y=yy, data=zdata) 
```

This way the fit object can link the keyword arguments to the corrosponding variables, and throw an error if no Variable by that name exists:

```python
fit = Fit(new, q=xx, y=yy, data=zdata) # raises TypeError: no variable by the name of 'q' exists.
```

Furthermore, it would still be possible to make the data a postional argument, but the variables would have to be keyword arguments:
```python
fit = Fit(new, zdata, x=xx, y=yy)
```

What do you think? Do you like one of these suggestions, or would you like something else? More ideas are welcome.

**Comment by Eljee** [comment]
Depending on the background of the experiment the data would be from, either the data or the model is most important. When a model is already known and the experiment should fit or be discarded, i.e. the fit is proof the experiment did not yield usable results. Or the experiment is very new and no model is proposed yet, making the data more important than the fit.

With that in mind, it should always be clear to people reading the fit what the data is and what the variables would be. (The example here is quite clear since the data is called data, this might not be the case for all examples or users.) Therefore I would say that the use of keyword-arguments is preferable. Note then, that this is a consideration solely based on the experimentalist in me, not the programmer.

**Comment by tBuLi** [comment]
For completeness, I'll paraphrase from the email of a friend. He prefers the second option because:
- model should be independent of the data.
- keyword argument allows for more dynamic code by dictionary unpacking (**kwargs).
- you could still accept a positional argument for the data if there is only one variable.

**Comment by tBuLi** [comment]
I wholeheartedly agree with the first two bullet points in the previous comment, but I'm not sure about the third, purely because internally this could seriously mess up the inheritance structure of fitting objects in symfit.

This discussion actually ties in nicely with my desire to support vector valued functions at some point. (https://github.com/tBuLi/symfit/issues/15) Because for the data, it would also be nice if you where free to use whatever keyword you prefer:

```python
x, y = variables('x, y')
z = x**2 + y**2

fit = Fit(model=z, x=xdata, y=ydata, z=zdata)
```
vector valued example:
```python
x, y = variables('x, y')
z_1 = x**2 + y**2
z_2 = 2 * (x + y)
z = z1, z2  # This line would depend on which vector syntax becomes accepted. See issue #15.

fit = Fit(model=z, x=xdata, y=ydata, z_1=z1data, z_2=z2data)
```

For this to work, a model would also need to have a name. If the syntax as provided here should literally work, we would have to use inspect upon assignment to check the name you desire, which I would prefer not to do. Of course adding the line ```z.name = 'z'``` is not too much overhead but it still looks ridiculous. This would make more sense:

```python
# Scalar function
z = Model(z=x**2 + y**2)
# Vector function
z = Model(z_1=x**2 + y**2, z_2=2 * (x + y))

fit = Fit(model=z, x=xdata, y=ydata, z_1=z1data, z_2=z2data)
```

This way would be pretty much perfect, because now all the naming of components of the vector valued function is completely transparent and unambiguous.

There are only two drawbacks I can think of:
- these changes would break backwards compatibility of the API, forcing users to change their code. But since this would make debugging so much easier, I think it's a change people won't mind, and this is still a young project.
- In case of a simple scalar function, I would still prefer the current way of doing things: ```z = x**2 + y**2```. It just looks a lot cleaner than putting ```Model``` in everywhere. So I would prefer not to loose this way of doing things. Maybe a little inspect isn't bad?

Any thoughts?

**Comment by tBuLi** [comment]
I have started work on implementing the keyword-style, e.g.
```python
fit = Fit(new, x=xdata, y=ydata, data=zdata) 
```
Named dependent variables are not supported yet, for the moment I will stick to scalar valued functions whose data *has* to be supplied via the ```data``` keyword.

Positional arguments will also be supported. The order is given as model, variables in alphabetical order, data, optional arguments for the ```Fit``` object at hand, e.g. sigma. So in the above line, the following would be equivalent:
```python
fit = Fit(new, xdata, ydata, zdata) 
```
This guarantees backwards compatibility in the case of 1 scalar to 1 scalar functions, which will be most use cases.

Of course the recommended way is by using keyword arguments, as this increases readability enormously. This will be the only way thought in the docs.

By the way, I posted a new suggestion for vector valued functions in #15. This new syntax might be the best to date.
